### ISSUE TITLE
Cyclic object asserts sometimes recurse infinitely

### ISSUE THREAD

**Issue by sparkprime** [google/jsonnet#451]
Simple example of a cyclic assert that logically should be OK:

```
{
    A: {
        assert self.f == $.B.f,
        f: 1,
    },  
    B: {
        assert self.f == $.A.f,
        f: 1,
    },  
}
```

What's happening here is in the assert statement, $.B is generating a new object, and accessing .g of that object triggers a fresh set of assertion evaluations.

Factoring out the $.A does not help:

```
{
    local a = self.A,
    local b = self.B,
    A: {
        assert self.f == b.f,
        f: 1,
    },  
    B: {
        assert self.f == a.f,
        f: 1,
    },
}   
```

However leaning on the mutually recursive local is a work around:

```
local
  A = {
    assert self.f == B.f,
    f: 1,
  },
  B = {
    assert self.f == A.f,
    f: 1,
  };
  
B 
```

And this can be done inside the object too:

```
{
    local outer = self,
    local A = {
        assert self.f == B.f,
        f: 1,
    },  
    local B = {
        assert self.f == A.f,
        f: 1,
    },
    A: A,
    B: B,
}   
```

However, these kinds of refactorings ought not to affect the termination behavior of Jsonnet. What is happening is that buried in the code there is a trick to break the assertion cycle if an object is already having its assertions evaluated. However that test uses the reference of the object, which is not valid in a functional language.

An even more trivial example is:

```
{
    assert self.f == 1,  // works
    assert (self {}).f == 1,  // doesn't work
    f: 1,
}
```

This problem has been observed in a real config written by @rossmomax.

**Comment by sparkprime** [comment]
Simplest solution is to disable all object assert evaluation when you're already in an assert, but those semantics may be surprising because it means that code that works fine inside an object assert can't necessarily be executed outside of it without getting an error.

**Comment by sbarzowski** [comment]
> Simplest solution is to disable all object assert evaluation when you're already in an assert, but those semantics may be surprising because it means that code that works fine inside an object assert can't necessarily be executed outside of it without getting an error.

It would affect semantics in a really strange way. Consider:
```
local errorObj = {assert false, f: 42};
local f = errorObj.f;
{
  assert f == 42
  f: f
}
```
It would finish without error under proposed semantics, if I understand the idea correctly.

**Comment by sparkprime** [comment]
Yes that is my proposal, but it's the only thing I can think of that will have consistent semantics other than removing object assertions completely.

Requiring cacheing isn't a good idea because you can't control memory use anymore.

**Comment by sparkprime** [comment]
We do essentially require cacheing for mutually recursive local variables though.

**Comment by sbarzowski** [comment]
I replied about caching in another issue, it seems a bit off-topic here. 

Link: https://github.com/google/go-jsonnet/issues/113

**Comment by sparkprime** [comment]
It doesn't help for the last example though, right?

**Comment by sbarzowski** [comment]
No, but it can be very easily modified so it does:
```
{
    local next = self {}
    assert self.f == 1,  // works
    assert next.f == 1,  // doesn't work
    f: 1,
}
```
With cached locals it's easy to add caching when it's needed to break such endless chains. And it may be enough - it will be always possible to write a nonterminating assert (halting problem etc.), so what else do we have to gain?

**Comment by sparkprime** [comment]
I think that will still not terminate because `self { }` will create a new object for which there is an empty cache

**Comment by sparkprime** [comment]
Here is a legitimate non-terminating case:

```
local M = { f+: 1 };
{
    assert (self + M).f < 10,
    f: 0,
}
```

**Comment by sparkprime** [comment]
What I don't like is that until now, `e + { }` was always an identity operation.

**Comment by sparkprime** [comment]
Another take on this is that it's like a desugaring that adds a boolean field and requires it to be true on all other fields. The problem is that this creates self references that would usually not be allowed:

`{ theAsserts: self.f < 10, f: assert self.theAsserts; 0, }`

**Comment by sparkprime** [comment]
I think any kind of field cacheing would also break the idempotency of + {}, since that can reset the cache and cause non-termination where there was previously termination. This is not the case for local vars since they are static and you cannot "make a new one" dynamically, with a fresh cache.

**Comment by sbarzowski** [comment]
You convinced me that independently from what is decided about object caching, some fixes are needed for assertion semantics. The dependence of termination on whether some stuff is cached or not is the problem. 

> Another take on this is that it's like a desugaring that adds a boolean field and requires it to be true on all other fields. The problem is that this creates self references that would usually not be allowed: `{ theAsserts: self.f < 10, f: assert self.theAsserts; 0, }`

Yeah, what current implementations are doing is technically inconsistent with official semantics. It's not allowed to index a field before previously checking that assertions hold according to `object-index`.  Our "yeah, but we're going to check assertions for this objects anyway later, so we can skip it for now" is technically a "proof by assuming the thesis". 

To solve it we must come up with better semantics. Here's my attempt:

self-without-assertions. Referring to field through self (or super) won't trigger assertions. If the object was referred from outside (not through self) the assertions will still run. This is very similar in practice to how it currently works - it follows similar logic "we're already inside the object so assertions are going to be checked anyway" - but it doesn't depend on the reference of the object.

The downside is that the set of nonterminating programs will be bigger than now. In particular it means that (even with caching) the examples from the issue will not terminate. To let the users combat this, we would need some alternative way to access object fields from outside which doesn't check the assertions (something else than `.`, for example it could be `std.accessField` or new construct `obj--->field`).

Please let me know what you think about this idea.

> I think that will still not terminate because self { } will create a new object for which there is an empty cache.

Right. It's will create endless chain of deeper and deeper objects. We're referring to any object only once, so caching doesn't do a thing.

> `e + { }` was always an identity operation.

In terms of semantics - it was and it would be. This is just evaluation strategy. In terms of performance it is significant and would be even more with caching. Caching doesn't change anything in terms of official semantics. It only affects performance and has a funny interaction with the way we bypass the semantics of assertions.

> I think any kind of field cacheing would also break the idempotency of + {}, since that can reset the cache and cause non-termination where there was previously termination. 

Same as above. It can only flip non-termination if it is already non-terminating according to formal semantics, if I understand `object-index` rule correctly. Otherwise it only affects performance.

> This is not the case for local vars since they are static and you cannot "make a new one" dynamically, with a fresh cache.

Interesting point. It's not exactly the same, but in a way you can achieve the same result with functions and arrays or locals:

```
local refreshable(f) = [f(), function() refreshable(f)];
local getValue(x) = x[0];
local refresh(x) = x[1]();

local foo = refreshable(function() range(1, 1000));
local bar = refresh(foo);
[
    getValue(foo), // evaluation needed
    getValue(foo), // cached
    getValue(refresh(foo)), // evaluation needed
    getValue(refresh(foo)), // evaluation needed
    getValue(bar), // evaluation needed
    getValue(bar), // cached
]
```
